Add example for collocation of transformer and predictor (#257)
kserve/website: docs/modelserving/v1beta1/transformer/collocation/README.md

# Collocate transformer and predictor in same pod

> KServe by default deploys the Transformer and Predictor as separate services so, you can deploy them on different devices and scale independently.

Reviewer: ```suggestion
KServe by default deploys the Transformer and Predictor as separate services, allowing you to deploy them on different devices and scale them independently.
```

Author: done

> <br/>However, In below scenarios you can choose to collocate the transformer and predictor in the same pod.

Reviewer: ```suggestion
<br/> Nevertheless, there are certain situations where you might prefer to collocate the transformer and predictor within the same pod. Here are a few scenarios:
```

Author: done

> 1. If your transformer is tightly coupled with the predictor and you want to do canary together.

Reviewer: ```suggestion
1. If your transformer is tightly coupled with the predictor and you want to perform canary deployment together.
```

Author: done

2. If you want to reduce sidecar resources.
3. If you want to reduce networking latency.

## Before you begin 

1. Your ~/.kube/config should point to a cluster with [KServe installed](../../../../get_started/README.md#install-the-kserve-quickstart-environment).
2. Your cluster's Istio Ingress gateway must be [network accessible](https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-control/).
3. You can find the [code samples](https://github.com/kserve/kserve/tree/master/docs/samples/v1beta1/transformer/collocation) on kserve repository.

## Deploy the InferenceService

Since, the predictor and the transformer are in the same pod, they need to listen on different ports to avoid conflict. `Transformer` is configured to listen on port 8000 and 8081 
while, `Predictor` listens on port 8080 and 8082. `Transformer` calls `Predictor` on port 8082 via local socket. 
Deploy the `Inferenceservice` using the below command.

```bash
cat <<EOF | kubectl apply -f -
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: custom-transformer-collocation
spec:
  predictor:
    containers:

>       - name: kserve-container

Reviewer: We probably need to add a note for the limitation that currently you need to use the custom container spec for kserve model container, we can create an issue to support model spec like the normal case.

Author: done

Author: issue created https://github.com/kserve/kserve/issues/2909

        image: kserve/custom-model-grpc:latest
        args:
          - --model_name=custom-model
          - --grpc_port=8082
          - --http_port=8080

      - image: kserve/image-transformer:latest
        name: transformer-container    # Do not change the container name
        args:
          - --model_name=custom-model
          - --protocol=grpc-v2
          - --http_port=8000
          - --grpc_port=8081
          - --predictor_host=localhost:8082
        ports:
          - containerPort: 8000
            protocol: TCP
EOF
```
!!! success "Expected output"
    ```{ .bash .no-copy }
    $ inferenceservice.serving.kserve.io/custom-transformer-collocation created
    ```
!!! Warning
    Always use the transformer container name as `transformer-container`. Otherwise, the model volume is not mounted to the transformer container which may result in an error.

> ## Check InferenceService status

Reviewer: ```suggestion

## Check InferenceService status
```

Author: done

```bash
kubectl get isvc custom-transformer-collocation
```
!!! success "Expected output"
    ```{ .bash .no-copy }
    NAME                             URL                                                         READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                              AGE
    custom-transformer-collocation   http://custom-transformer-collocation.default.example.com   True           100                              custom-transformer-collocation-predictor-00001   133m
    ```

> !!! Note

Reviewer: ```suggestion

!!! Note
```

Author: done

    If your DNS contains `svc.cluster.local`, then `Inferenceservice` is not exposed through Ingress. you need to [configure DNS](https://knative.dev/docs/install/yaml-install/serving/install-serving-with-yaml/#configure-dns) 
    or [use a custom domain](https://knative.dev/docs/serving/using-a-custom-domain/) in order to expose the `isvc`.

## Run a prediction
Prepare the inputs for the inference request. Copy the following Json into a file named `input.json`.

> ```json

Reviewer: This is a bit large to add to the page itself, we can link it and allow user to download.

Author: done
